[Dynamic RPC] Allow existing ranks to communicate with newly joined ranks

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #75811 [RPC small change] Improve logging from 'unknown destination worker'
* #74561 [Dynamic RPC] Add graceful shutdown for dynamic RPC members
* **#74035 [Dynamic RPC] Allow existing ranks to communicate with newly joined ranks**

This PR allows for existing ranks in Dynamic RPC to communicate with ranks that have newly joined. That is, rank N will be able to run RPC against all ranks >= N + 1. Previously only new ranks can talk to existing ranks, now this PR enables all ranks to communicate with each other.

This includes support for CUDA rpc. Device maps and reverse device maps are aggregated and exchanged.

Previously:

Process 1 (init and send would fail):
```python
init_rpc("worker0", rank=0)
# send to worker1 which joins after worker0
rpc.sync("worker1", torch.add, (torch.tensor(1), torch.tensor(1)))
```

Process2 (command against a rank that already joined, would fail):
```python
# joins after worker0
init_rpc("worker1", rank=1)
```

Now:
Above scenario succeeds

TODO:
- Have worker block with timeout if the worker does not exist in its name-to-id mappings (add in future PR)

H-Huang committed:
[Dynamic RPC] Allow existing ranks to communicate with newly joined ranks

[ghstack-poisoned]

H-Huang committed:
Update on "[Dynamic RPC] Allow existing ranks to communicate with newly joined ranks"


WIP

[ghstack-poisoned]

H-Huang committed:
Update on "[Dynamic RPC] Allow existing ranks to communicate with newly joined ranks"


WIP

[ghstack-poisoned]

H-Huang committed:
Update on "[Dynamic RPC] Allow existing ranks to communicate with newly joined ranks"


WIP

[ghstack-poisoned]

H-Huang committed:
Update on "[Dynamic RPC] Allow existing ranks to communicate with newly joined ranks"


WIP

[ghstack-poisoned]

H-Huang committed:
Update on "[Dynamic RPC] Allow existing ranks to communicate with newly joined ranks"


WIP

[ghstack-poisoned]

H-Huang committed:
Update on "[Dynamic RPC] Allow existing ranks to communicate with newly joined ranks"


WIP

[ghstack-poisoned]

mrshenli commented on torch/_C/_distributed_rpc.pyi:
```diff
@@ -100,6 +100,7 @@ class TensorPipeAgent(RpcAgent):
     def get_worker_info(self, id: int) -> WorkerInfo: ...
     def get_worker_infos(self) -> List[WorkerInfo]: ...
     def _get_device_map(self, dst: WorkerInfo) -> Dict[torch.device, torch.device]: ...
+    def _update_group_membership(self): ...
```
do we need to add `_get_backend_options` here as well?

mrshenli commented on torch/distributed/rpc/backend_registry.py:
```diff
@@ -267,6 +273,56 @@ def _tensorpipe_check_local_device_maps(name, device_count, options):
+
+def _get_device_infos():
+    if torch.cuda.is_available():
+        device_count = torch.cuda.device_count()
```
`torch.cuda.device_count()` should work even CUDA is not available. 

https://github.com/pytorch/pytorch/blob/cd80ef6cdbc413dbe6a22afd55101168306f7020/torch/cuda/__init__.py#L453-L458

mrshenli commented on torch/distributed/rpc/backend_registry.py:
```diff
@@ -267,6 +273,56 @@ def _tensorpipe_check_local_device_maps(name, device_count, options):
+        device_count = 0
+    agent = api._get_current_rpc_agent()
+    opts = agent._get_backend_options()
+    if torch.cuda.is_available():
+        device_count = torch.cuda.device_count()
+    else:
+        device_count = 0
```
duplicated code

mrshenli commented on torch/distributed/rpc/backend_registry.py:
```diff
@@ -356,14 +412,16 @@ def _tensorpipe_init_backend_handler(store, name, rank, world_size, rpc_backend_
                     pass
                 except Exception:
-                    api.shutdown()
+                    # api.shutdown()
```
what exceptions are we expecting here?

H-Huang replied:
I will remove this exception since we dont have shutdown implementation at the moment.

mrshenli commented on torch/distributed/rpc/backend_registry.py:
```diff
@@ -356,14 +412,16 @@ def _tensorpipe_init_backend_handler(store, name, rank, world_size, rpc_backend_
+                    _set_devices_and_reverse_device_map(agent, rank, rpc_backend_options.device_maps)
                     pass
                 except Exception:
-                    api.shutdown()
```
I assume we need a non-graceful shutdown here?

H-Huang replied:
I will add non-graceful shutdown in next PR!

mrshenli commented on torch/distributed/rpc/backend_registry.py:
```diff
@@ -267,6 +273,56 @@ def _tensorpipe_check_local_device_maps(name, device_count, options):
+    print(f"{my_rank}: _set_devices_and_reverse_device_map, {my_device_maps}")
+    my_worker_info = agent.get_worker_info()
+    my_name = my_worker_info.name
+    all_worker_infos = agent.get_worker_infos()
```
Just to double check, this is retrieved from store, right?

H-Huang replied:
It retrieved from agent's local cache, but ultimately retrieved from store. On TPagent initialization, it retrieves all current workers from store and saves it in a map. `agent.get_worker_infos()` just returns that map. Since this method is called directly after initialization and the group membership is locked at this time, the state of the workers is valid

mrshenli commented on torch/distributed/rpc/backend_registry.py:
```diff
@@ -267,6 +273,56 @@ def _tensorpipe_check_local_device_maps(name, device_count, options):
+        device_count = 0
+    return device_count, opts.device_maps, opts.devices
+
+def _set_devices_and_reverse_device_map(agent, my_rank, my_device_maps):
```
`my_device_maps` not used in this function? 

Just wanna make sure I understand the logic correctly. If `my_device_maps` contains a worker X that's not joined yet, that part of my device map will be activated and verified when that future worker X calls this method?

H-Huang replied:
Removed `my_device_maps`.

>Just wanna make sure I understand the logic correctly. If my_device_maps contains a worker X that's not joined yet, that part of my device map will be activated and verified when that future worker X calls this method?

Yes, is an existing worker has a worker not yet joined but in it's device map (in your example "worker X"), the devices of worker X is not yet checked. When worker X joins, the device map of itself and the workers targeting it are validated.

mrshenli commented on torch/distributed/rpc/backend_registry.py:
```diff
@@ -240,8 +238,16 @@ def _tensorpipe_exchange_and_check_all_device_maps(
+    for node in all_names:
+        if my_name in all_device_maps[node]:
+            reverse_device_maps[node] = {
+                v: k for k, v in all_device_maps[node][my_name].items()
```
need to verify this device is valid on this current node?

H-Huang replied:
Added validation logic for the device maps in func `_validate_device_maps()`

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -1154,6 +1154,91 @@ std::vector<WorkerInfo> TensorPipeAgent::getWorkerInfos() const {
+    //     }
+    //   }
+    // }
+    reverseDeviceMaps_ = reverseDeviceMaps;
```
curious, any reason that we need to update the entire `reverseDeviceMaps_ `, instead of just insert the info for the newly joined node? If this new node's info already exist in the `reverseDeviceMaps_`, it should return an error?

H-Huang replied:
Every time a new worker joins I am performing validation of all the device maps currently set by the group's workers. I am also constructing the reverseDeviceMap so I just passed it in here for simplicity. Now I updated the function to add the map only if the worker does not exist, in case of the new node info already existing, that error will be checked and thrown in `_validate_device_maps()`.

mrshenli reviewed (commented):
The high-level structure LGTM! Thanks for putting this together!

H-Huang committed:
Update on "[Dynamic RPC] Allow existing ranks to communicate with newly joined ranks"


This PR allows for existing ranks in Dynamic RPC to communicate with ranks that have newly joined. That is, rank N will be able to run RPC against all ranks >= N + 1. Previously only new ranks can talk to existing ranks, now this PR enables all ranks to communicate with each other.

This includes support for CUDA rpc. Device maps and reverse device maps are aggregated and exchanged.

Previously:

Process 1 (init and send would fail):
```python
init_rpc("worker0", rank=0)
# send to worker1 which joins after worker0
rpc.sync("worker1", torch.add, (torch.tensor(1), torch.tensor(1)))
```

Process2 (command against a rank that already joined, would fail):
```python
# joins after worker0
init_rpc("worker1", rank=1)
```

Now:
Above scenario succeeds

TODO:
- Have worker block with timeout if the worker does not exist in its name-to-id mappings
- Test case for this scenario



[ghstack-poisoned]

H-Huang committed:
Update on "[Dynamic RPC] Allow existing ranks to communicate with newly joined ranks"


This PR allows for existing ranks in Dynamic RPC to communicate with ranks that have newly joined. That is, rank N will be able to run RPC against all ranks >= N + 1. Previously only new ranks can talk to existing ranks, now this PR enables all ranks to communicate with each other.

This includes support for CUDA rpc. Device maps and reverse device maps are aggregated and exchanged.

Previously:

Process 1 (init and send would fail):
```python
init_rpc("worker0", rank=0)
# send to worker1 which joins after worker0
rpc.sync("worker1", torch.add, (torch.tensor(1), torch.tensor(1)))
```

Process2 (command against a rank that already joined, would fail):
```python
# joins after worker0
init_rpc("worker1", rank=1)
```

Now:
Above scenario succeeds

TODO:
- Have worker block with timeout if the worker does not exist in its name-to-id mappings
- Test case for this scenario



[ghstack-poisoned]

H-Huang committed:
Update on "[Dynamic RPC] Allow existing ranks to communicate with newly joined ranks"


This PR allows for existing ranks in Dynamic RPC to communicate with ranks that have newly joined. That is, rank N will be able to run RPC against all ranks >= N + 1. Previously only new ranks can talk to existing ranks, now this PR enables all ranks to communicate with each other.

This includes support for CUDA rpc. Device maps and reverse device maps are aggregated and exchanged.

Previously:

Process 1 (init and send would fail):
```python
init_rpc("worker0", rank=0)
# send to worker1 which joins after worker0
rpc.sync("worker1", torch.add, (torch.tensor(1), torch.tensor(1)))
```

Process2 (command against a rank that already joined, would fail):
```python
# joins after worker0
init_rpc("worker1", rank=1)
```

Now:
Above scenario succeeds

TODO:
- Have worker block with timeout if the worker does not exist in its name-to-id mappings
- Test case for this scenario



[ghstack-poisoned]

H-Huang committed:
Update on "[Dynamic RPC] Allow existing ranks to communicate with newly joined ranks"


This PR allows for existing ranks in Dynamic RPC to communicate with ranks that have newly joined. That is, rank N will be able to run RPC against all ranks >= N + 1. Previously only new ranks can talk to existing ranks, now this PR enables all ranks to communicate with each other.

This includes support for CUDA rpc. Device maps and reverse device maps are aggregated and exchanged.

Previously:

Process 1 (init and send would fail):
```python
init_rpc("worker0", rank=0)
# send to worker1 which joins after worker0
rpc.sync("worker1", torch.add, (torch.tensor(1), torch.tensor(1)))
```

Process2 (command against a rank that already joined, would fail):
```python
# joins after worker0
init_rpc("worker1", rank=1)
```

Now:
Above scenario succeeds

TODO:
- Have worker block with timeout if the worker does not exist in its name-to-id mappings (add in future PR)


[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -561,7 +561,13 @@ void TensorPipeAgent::pipeRead(
+    if (isStaticGroup_) {
+      streams = getStreamsFromPoolForDevices(devices_);
+    } else {
+      std::unique_lock<std::mutex> lock(groupMembershipMutex_);
```
nit: since we don't need to lock and unlock here, a `lock_guard` might be better?

H-Huang replied:
Switched it with custom lock guard as suggested

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -641,6 +647,7 @@ void TensorPipeAgent::sendCompletedResponseMessage(
 
     for (const auto& tensor : responseMessage->tensors()) {
       const auto device = tensor.device();
+      std::unique_lock<std::mutex> lock(groupMembershipMutex_);
```
why do we need this lock here? and why this is inside the for loop? Since this is on the critical path for every RPC response, will this cause any perf issues?

H-Huang replied:
Thanks for catching. This was meant to also have the if statement check to only run for dynamic rpc, added the lock guard to this as well

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -1156,12 +1188,55 @@ std::vector<WorkerInfo> TensorPipeAgent::getWorkerInfos() const {
 const std::string& TensorPipeAgent::findWorkerURL(
     const WorkerInfo& worker) const {
-  const auto it = workerNameToURL_.find(worker.name_);
+  std::unordered_map<std::string, std::string>::const_iterator it;
+  if (isStaticGroup_) {
+    it = workerNameToURL_.find(worker.name_);
+  } else {
+    std::unique_lock<std::mutex> lock(groupMembershipMutex_);
+    it = workerNameToURL_.find(worker.name_);
+  }
```
nit: since this code pattern has occurred multiple times, it might worth create a custom guard for it? E.g.

```c++
{
  SomeGuard guard(groupMembershipMutex_);
  it = workerNameToURL_.find(worker.name_);
}
```

H-Huang replied:
That is a great suggestion, thanks!

H-Huang committed:
Update on "[Dynamic RPC] Allow existing ranks to communicate with newly joined ranks"


This PR allows for existing ranks in Dynamic RPC to communicate with ranks that have newly joined. That is, rank N will be able to run RPC against all ranks >= N + 1. Previously only new ranks can talk to existing ranks, now this PR enables all ranks to communicate with each other.

This includes support for CUDA rpc. Device maps and reverse device maps are aggregated and exchanged.

Previously:

Process 1 (init and send would fail):
```python
init_rpc("worker0", rank=0)
# send to worker1 which joins after worker0
rpc.sync("worker1", torch.add, (torch.tensor(1), torch.tensor(1)))
```

Process2 (command against a rank that already joined, would fail):
```python
# joins after worker0
init_rpc("worker1", rank=1)
```

Now:
Above scenario succeeds

TODO:
- Have worker block with timeout if the worker does not exist in its name-to-id mappings (add in future PR)


[ghstack-poisoned]

H-Huang committed:
Update on "[Dynamic RPC] Allow existing ranks to communicate with newly joined ranks"


This PR allows for existing ranks in Dynamic RPC to communicate with ranks that have newly joined. That is, rank N will be able to run RPC against all ranks >= N + 1. Previously only new ranks can talk to existing ranks, now this PR enables all ranks to communicate with each other.

This includes support for CUDA rpc. Device maps and reverse device maps are aggregated and exchanged.

Previously:

Process 1 (init and send would fail):
```python
init_rpc("worker0", rank=0)
# send to worker1 which joins after worker0
rpc.sync("worker1", torch.add, (torch.tensor(1), torch.tensor(1)))
```

Process2 (command against a rank that already joined, would fail):
```python
# joins after worker0
init_rpc("worker1", rank=1)
```

Now:
Above scenario succeeds

TODO:
- Have worker block with timeout if the worker does not exist in its name-to-id mappings (add in future PR)


[ghstack-poisoned]

mrshenli reviewed (approved):
Thanks @H-Huang. LGTM

mrshenli commented:
Hey @H-Huang, could you please import this PR to check it does not block any internal use cases? Thanks!

H-Huang commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -4377,6 +4377,99 @@ def test_without_world_size_new_rank_can_communicated_with_existing_rank(self):
+    # Dynamic RPC existing ranks can communicate with new ranks
+    @dist_init(setup_rpc=False)
+    def test_without_world_size_existing_rank_can_communicate_with_new_rank(self):
+        # TODO: Using process group for synchronization to ensure rank 0 is created first
```
Yep! For TCPStore the ranks != 0 will block waiting for master store to be created on rank 0. For filestore, none of the ranks block, but for this test case I wanted to ensure that rank 0 was created first to test new ranks communicating with existing rank.

H-Huang commented:
I am not able to import this PR due to github first being out of sync with master.

H-Huang commented:
@pytorchbot merge this
